tests/std/tests/P0912R5_coroutine/test.cpp

```diff
@@ -0,0 +1,127 @@
+
+#include <version> // TRANSITION, P0912R5 Library Support For Coroutines
+#if defined(__cpp_lib_coroutine) && __cpp_lib_coroutine >= 201902L // TRANSITION, P0912R5 Library Support For Coroutines
+
+#include <assert.h>
+#include <coroutine>
+#include <exception>
+using namespace std;
+
+int g_tasks_destroyed{0};
+
+struct Task {
+    struct Promise {
+        int result{-1000};
+        coroutine_handle<> previous;
+
+        Task get_return_object() {
+            return {*this};
+        }
+
+        suspend_always initial_suspend() {
+            return {};
+        }
+
+        auto final_suspend() noexcept {
+            struct Awaiter {
+                bool await_ready() noexcept {
+                    return false;
+                }
+
+                void await_resume() noexcept {}
+
+                coroutine_handle<> await_suspend(coroutine_handle<Promise> h) {
+                    if (auto& pre = h.promise().previous; pre) {
+                        return pre; // resume awaiting coroutine
+                    }
+
+                    // If there is no previous coroutine to resume, we've reached the outermost coroutine.
+                    // Return a noop coroutine to allow control to return back to the caller.
+                    return noop_coroutine();
+                }
+            };
+
+            return Awaiter{};
+        }
+
+        void return_value(const int v) {
+            result = v;
+        }
+
+        void unhandled_exception() noexcept {
+            terminate();
+        }
+    };
+
+    using promise_type = Promise;
+
+    bool await_ready() const noexcept {
+        return false;
+    }
+
+    int await_resume() {
+        return coro.promise().result;
+    }
+
+    auto await_suspend(coroutine_handle<> enclosing) {
+        coro.promise().previous = enclosing;
+        return coro; // resume ourselves
+    }
+
+    Task(Task&& rhs) noexcept : coro(rhs.coro) {
+        rhs.coro = nullptr;
+    }
+
+    Task(Promise& p) : coro(coroutine_handle<Promise>::from_promise(p)) {}
+
+    ~Task() {
+        ++g_tasks_destroyed;
+
+        if (coro) {
+            coro.destroy();
+        }
+    }
+
+    coroutine_handle<Promise> coro;
+};
+
+Task triangular_number(const int n) {
+    if (n == 0) {
+        co_return 0;
+    }
+
+    co_return n + co_await triangular_number(n - 1);
+}
+
+int main() {
+    assert(g_tasks_destroyed == 0);
+
+    {
+        Task t               = triangular_number(10);
+        coroutine_handle<> h = t.coro;
+
+        assert(h == t.coro);
+        assert(h);
+        assert(!h.done());
+
+        h();
+
+        assert(h == t.coro);
+        assert(h);
+        assert(h.done());
+
+        assert(g_tasks_destroyed == 10); // triangular_number() called for [0, 9]
+
+        const int val = t.coro.promise().result;
+
+        assert(val == 55);
+    }
+
+    assert(g_tasks_destroyed == 11); // triangular_number() called for [0, 10]
+}

```

Comment:
This functionality test falls far short of validating all the members of three different specializations of `coroutine_handle`. Should it do so? Should we have a separate test?

Reply:
My preference is to merge this fairly simple end-to-end test, and file a followup `test` issue for enhancing coverage (which will be easier when the necessary compiler is publicly available).